vunit/vc/compliance_test.py

```diff
@@ -0,0 +1,339 @@
+            test.add_config(
+                name="fail_on_unexpected_msg_type=%s"
+                % str(fail_on_unexpected_msg_type),
+                generics=dict(
+                    fail_on_unexpected_msg_type=fail_on_unexpected_msg_type,
+                    use_custom_logger=True,
+                    use_custom_actor=True,
+                ),
+            )
+
+        return tb_file
+
+    def create_vhdl_testbench(self):
+        """Returns a VHDL compliance testbench."""
+
+        entity_name = self.vc_entity.identifier
+
+        library_names = set()
+        context_items = "library %s;\n" % self.vc_lib.name
+        context_items += "use %s.%s.all;\n" % (self.vc_lib.name, self.vci_name)
+        for ref in self.vc_code.references:
+            if ref.is_package_reference() or ref.is_context_reference():
+                if ref.library_name != "work":
+                    if ref.library_name not in library_names:
+                        library_names.add(ref.library_name)
+                        context_items = (
+                            "library %s;\n" % ref.library_name + context_items
+                        )
+
+                library_name = (
+                    ref.library_name if ref.library_name != "work" else self.vc_lib.name
+                )
+
+                if ref.is_context_reference():
+                    context_items += "context %s.%s;\n" % (
+                        library_name,
+                        ref.design_unit_name,
+                    )
+
+                if ref.is_package_reference():
+                    context_items += "use %s.%s.%s;\n" % (
+                        library_name,
+                        ref.design_unit_name,
+                        ref.name_within,
+                    )
+
+        signal_declarations = ""
+        port_mappings = ""
+        for port in self.vc_entity.ports:
+            if (port.mode == "in" or port.mode == "inout") and port.init_value is None:
+                signal_declarations += "  signal %s : %s;\n" % (
+                    ", ".join(port.identifier_list),
+                    port.subtype_indication,
+                )
+                for identifier in port.identifier_list:
+                    port_mappings += "      %s => %s,\n" % (identifier, identifier)
+            else:
+                for identifier in port.identifier_list:
+                    port_mappings += "      %s => open,\n" % identifier
+
+        vc_instantiation = """  vc_inst: entity %s.%s
+    generic map(%s);
+""" % (
+            self.vc_lib.name,
+            entity_name,
+            self.vc_entity.generics[0].identifier_list[0],
+        )
+
+        if len(self.vc_entity.ports) > 0:
+            vc_instantiation = (
+                vc_instantiation[:-2]
+                + """
+    port map(
+"""
+            )
+
+            vc_instantiation += port_mappings[:-2] + "\n    );\n"
+
+        tb_template = Template(
```

Comment:
Instead of 'hardcoding' the template in the Python source, would it be feasible to define it in a `*.vhd.tpl` template file?

Reply:
I agree that these code snippets are big and few enough to move elsewhere. However, I think your suggestion to do the same thing for `test_compliance__test.py` would result in too fragmented code. Those snippets are very small and moving them elsewhere would make the tests hard to read.